### PR TITLE
[Ryujit/ARM32] Implement NYI related with overflow for ARM

### PR DESCRIPTION
**Pull request by sjsinju**
On last comment of #8496, the NYI message of overflow checks is printed after running the CodeGenBringUpTests.

I tried to create the sample code for making same assertion.
```
using System;
namespace Checked
{
    class Program
    {   
        unsafe private void* m_value;
        public unsafe void Foo(long value)
        {
            m_value= (void*)checked((int)value);
        }
        static void Main(string[] args)
        {
            Program p = new Program();
            p.Foo(0x4234abcdL);
        }
    }   
}
```
It made same results.
```
Assert failure(PID 21515 [0x0000540b], Thread: 21515 [0x540b]): Assertion failed 'NYI_ARM: overflow checks' in 'System.IntPtr:.ctor(long):this' (IL size 11)
    File: /home/sjsujinkim/works/dotnet/coreclr/src/jit/lsraarm.cpp Line: 996
    Image: /home/sjsujinkim/works/dotnet/Windows_NT.x64.Release/Tests/coreoverlay/corerun
Aborted
```

That was the message about a temp register setup for overflow checks.

It was referenced
https://github.com/dotnet/coreclr/blob/master/src/jit/lsraarm64.cpp#L399

I think it doesn't make any problem even though writing it the same as arm64.

### CODE REVIEW

**Comment by hqueue** [comment]
@sjinju Please fix formatting failure in CI by applyign coding convention for JIT :)

**Comment by sjsinju** [comment]
@dotnet-bot test Please

**Comment by sjsinju** [comment]
@hqueue OK. It was changed.

**Comment by sjsinju** [comment]
If this codes are committed, follow messages are printed.
```
Assert failure(PID 32301 [0x00007e2d], Thread: 32301 [0x7e2d]): Assertion failed 'NYI: Unimplmented GT_CAST:int <--> int with overflow' in 'Checked.Program:Foo(long):this' (IL size 11)

    File: /home/sjsujinkim/works/dotnet/coreclr/src/jit/codegenarm.cpp Line: 615
    Image: /home/sjsujinkim/works/dotnet/Windows_NT.x64.Release/Tests/coreoverlay/corerun

Aborted
```
After merged, I can implement for another NYI of overflow.

**Comment by sjsinju** [comment]
Below is more additional related with overflow messages.
```
sjsujinkim@sjsujinkim:~/works/dotnet/coreclr/src$ grep -Rn NYI | grep -i overflow
jit/emitarm.cpp:7724:        NYI_ARM("emitInsTernary overflow");
jit/emitarm.cpp:7754:            NYI_ARM("emitInsTernary overflow");
jit/emitarm.cpp:7828:        NYI_ARM("emitInsTernary overflow");
jit/codegenarm.cpp:392:        NYI("Overflow not yet implemented");
jit/codegenarm.cpp:581:                NYI("Overflow not yet implemented");
jit/codegenarm.cpp:615:                NYI("Unimplmented GT_CAST:int <--> int with overflow");
jit/codegenarm.cpp:2424:        NYI("genLongToIntCast: overflow check");
jit/codegenarm.cpp:2486:        NYI_ARM("CodeGen::genIntToIntCast for OverflowCheck");
```

**Comment by BruceForstall** [comment]
cc @dotnet/jit-contrib

**Comment by sjsinju** [comment]
I have checked the working of sample app is well by 4 commits until now.
```
sjsujinkim@sjsujinkim-rpi3:~/works/dotnet/tests/Checked/bin/Debug/netcoreapp1.1$ /home/sjsujinkim/works/dotnet/Windows_NT.x64.Release/Tests/coreoverlay/corerun Checked.dll
sjsujinkim@sjsujinkim-rpi3:~/works/dotnet/tests/Checked/bin/Debug/netcoreapp1.1$
```
And I changed the input value of Foo().
```
//p.Foo(0x4234abcdL);
p.Foo(0x8234abcdL);
```
I met below exeption
```
sjsujinkim@sjsujinkim-rpi3:~/works/dotnet/tests/Checked/bin/Debug/netcoreapp1.1$ /home/sjsujinkim/works/dotnet/Windows_NT.x64.Release/Tests/coreoverlay/corerun Checked.dll

Unhandled Exception: System.OverflowException: Arithmetic operation resulted in an overflow.
   at Checked.Program.Foo(Int64 value) in /home/sjsujinkim/works/dotnet/tests/Checked/Program.cs:line 12
   at Checked.Program.Main(String[] args) in /home/sjsujinkim/works/dotnet/tests/Checked/Program.cs:line 19
Aborted
```

**Comment by sjsinju** [comment]
```
sjsujinkim@sjsujinkim-rpi3:~/works/dotnet/coreclr$ grep NYI /home/sjsujinkim/works/logs/test/log_ryujit_func |sed 's|^.* - NYI ||' |sed 's|^.*src/jit/\(.*\))|\1|' |sort |uniq -c |sort -nr
   5556 lsraarm.cpp:1364 - NYI: lowering struct var
    420 lsraarm.cpp:1353 - NYI: Unimplemented node type lclHeap
    264 morph.cpp:4610 - NYI_ARM: fgMorphMultiregStructArgs
    188 lsraarm.cpp:901 - NYI: lowering double
     40 lsraarm.cpp:900 - NYI: lowering struct
      1 codegenarm.cpp:1878 - NYI: genRegCopy
```

**Comment by sjsinju** [comment]
@dotnet/jit-contrib I think It is okay enough. Please review.

File: src/jit/emitarm.cpp [hunk 1]
```diff
@@ -1380,7 +1380,7 @@ emitter::insSize emitter::emitInsSize(insFormat insFmt)
 
 /*****************************************************************************
  *
- *  emitIns_valid_imm_for_add() returns true when the immediate 'imm'
+ *  emitins_valid_imm_for_add() returns true when the immediate 'imm'
```
Comment (hqueue) [inline src/jit/emitarm.cpp]: Please fix typo here.

File: src/jit/emitarm.cpp [hunk 2]
```diff
@@ -1396,6 +1396,20 @@ emitter::insSize emitter::emitInsSize(insFormat insFmt)
...
 /*****************************************************************************
  *
+ *  emitins_valid_imm_for_cmp() returns true if this 'imm'
+ *   can be encoded as a input operand to an cmp instruction.
+ */
+/*static*/ bool emitter::emitIns_valid_imm_for_cmp(int imm, insFlags flags)
+{
+    if (isModImmConst(imm)) // funky arm immediate
+        return true;
+    if (isModImmConst(-imm)) // funky arm immediate via sub
```
Comment (hqueue) [inline src/jit/emitarm.cpp]: Comment is misleading, not `sub`.

File: src/jit/codegenarm.cpp [hunk 3]
```diff
@@ -610,10 +610,6 @@ void CodeGen::genCodeForTreeNode(GenTreePtr treeNode)
             // Cast is never contained (?)
             noway_assert(targetReg != REG_NA);
 
-            // Overflow conversions from float/double --> int types go through helper calls.
```
Comment (hqueue) [inline src/jit/codegenarm.cpp]: I think we should leave this comment.

**Comment by briansull** [comment]
Looks Good

**Pull request merged**